### ISSUE TITLE
svelte-migrate svelte-5 - Type in `$state<HERE>` ?

### ISSUE THREAD

**Issue by jycouet** [sveltejs/kit#12831]
### Describe the problem

Today when I tried migrate, I had: 

```ts
// svelte 4
let status: TileStatus = "Loading";
// after migrate:
let status: TileStatus = $state("Loading");
```

But then, I have some type error:
```
This comparison appears to be unintentional because the types '"Loading"' and '"Error"' have no overlap.
```

### Describe the proposed solution

```ts
// svelte 4
let status: TileStatus = "Loading";
// after migrate:
let status = $state<TileStatus>("Loading");
```

### Importance

would make my life easier

**Comment by jycouet** [comment]
Thx for your links & context.
I'm not sure I understand your conclusion actually.
Do you mean that you want to have the `// error here` ? Or you don't want ?

In svelte 4, `status` was not narrowed down in initiallisation thx to the type setting. `: TileStatus`
In Svelte 5, setting the type on `status` is a bit useless as the narrow will happen after the function (rune) and will say only "Loading" is valid. 

So, after running `svelte-migrate` everyone will need to change this by hand?
```diff
- let status: TileStatus = $state("Loading"); 
+ let status = $state<TileStatus>("Loading");
```

_There are probably cases that I'm not aware of, it just a naive thought._

**Comment by brunnerh** [comment]
`// error here` just points out where you can potentially get an error like the one mentioned.

> In Svelte 5, setting the type on `status` is a bit useless as the narrow will happen after the function (rune) and will say only "Loading" is valid.

There should not be an issue, the types are compatible in that direction, i.e. `"Loading"` is assignable to `TileStatus`.

Thinking about this a bit more — please give a _full example_ where Svelte 4 code did _not_ cause an error but Svelte 5 code _does_. You can run into the same error in Svelte 4 as well, so I am curious how it is caused by a migration.

**Comment by jycouet** [comment]
Yes sure, here is the repo in svelte 4 https://github.com/jycouet/remult-project-svelte-4

```bash
npm i
npm run check
# all good

npx svelte-migrate svelte-5
npm i
npm run check
# you will see the example of the issue `/src/demo/Admin.svelte`
```

**Comment by paoloricciuti** [comment]
btw this should be moved to svelte repo

**Comment by brunnerh** [comment]
So it is about a comparison in a reactive statement/`$derived` as suspected.
```js
$: subtitle = status === "Error" ? "There seems to be an issue" : "";
$: message =
    status === "Success"
    ? "Remult Admin is an autogenerated UI for managing your data."
    : "Failed to connect to Remult server.";
```

In theory, those comparisons [should have the same problem](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAKglgGwgZWAQ2AVwM5QLxQBEymAxqRNtoVAD5ECiATkwPZM32EDqaTAdnH4BzTkQCS-AGasxhADKs0AEyGiA3AFgAUEmBRs6LNgBcsRCiM58RRSrWEt2nXoOYARsDjAkAGigAtpTYaMIQTgAkZtgeXj7QBIYY1nipjCzsNAD8RDAAFhBM0NgQEAG4wKxQ7tBo-FBwVJgQNGaEjjpRgcGhCTpQAwZWuKkExGQUVIT9gzmEAEplmAj6AILKAUINuHVQaJiVYfyFGBDKUACq4lAyTIF1oWpQIKyYd8oYaAB009qDUG0AGJoCznSpQUisfjHUj6cGLALLfQlJgAN0KPycOggAA8wOx9ABvAC+6iAA), but I guess the language tooling passes a transformed version to TS where the type cannot be narrowed via control flow analysis (if it were not transformed the variables would not even exist).

Maybe the language tooling could somehow be adjusted in such a way that for `$derived` the narrowing is also prevented.

**Comment by paoloricciuti** [comment]
I wonder how was it working in svelte 4...label also has no meaning for TS so it doesn't know that it would be run again possibly with a different value.

**Comment by brunnerh** [comment]
The transformed code by `svelte2tsx` apparently looks like this:
```js
let  subtitle = __sveltets_2_invalidate(() => status === "Error" ? "There seems to be an issue" : "");
let  message =
  __sveltets_2_invalidate(() => status === "Success"
	? "Remult Admin is an autogenerated UI for managing your data."
	: "Failed to connect to Remult server.");
```
If that code is used by the tooling it works, the comparison is in a function and as long as `status` is assigned somewhere, the narrowing will not happen.

The transformed code for the migrated code does not contain a function call, it is taken as is:
```js
let subtitle = $derived(status === "Error" ? "There seems to be an issue" : "");
let message =
  $derived(status === "Success"
	? "Remult Admin is an autogenerated UI for managing your data."
	: "Failed to connect to Remult server.");
```

A transform from `$derived` to `$derived.by` should already help here.

**Comment by jycouet** [comment]
It's "just" a type issue, correct?

I find it strange to go with `$derived.by` if you are in TS, but if you are in JS it's ok to use `$derived`

There is something wrong with `let status = $state<TileStatus>("Loading");` ?

**Comment by brunnerh** [comment]
> It's "just" a type issue, correct?
>
> I find it strange to go with `$derived.by` if you are in TS, but if you are in JS it's ok to use `$derived`

More or less, the problem with `$derived` is that it's not a function but a compiler instruction. The code that is visible is not the code that will run which of course has the potential to confuse tooling when the code is used as is.

> There is something wrong with `let status = $state<TileStatus>("Loading");` ?

Yes, it's noisy due to extra punctuation and people generally seem to prefer to have the type at the variable declaration rather than the right side of the assignment. I have even seen some confusion over this syntax, likely because of unfamiliarity with generics.

**Comment by jycouet** [comment]
Thank you for giving these hints. You are right, I'm probably too familliar with `<generics>` that I don't see them ^^

What should/could I do to help ?

**Comment by brunnerh** [comment]
Unless my analysis is off (which is entirely possible), [`svelte2tsx`](https://github.com/sveltejs/language-tools/tree/master/packages/svelte2tsx) would need to be adjusted to transform `$derived` into the form `some_function(() => <derived_contents>)`. I have never touched that package though and don't know much about it.

**Comment by adiguba** [comment]
> Unless my analysis is off (which is entirely possible), [`svelte2tsx`](https://github.com/sveltejs/language-tools/tree/master/packages/svelte2tsx) would need to be adjusted to transform `$derived` into the form `some_function(() => <derived_contents>)`. I have never touched that package though and don't know much about it.

Yes, it seems that this is the problem.
=> TypeScript does not know that $derived() is actually a dynamic function, so it generates an error because at this stage the value cannot be different from its initialization.

It's equivalent to this, which can never be true :
```ts
    let status: TileStatus = "Loading";
    const hasError = status === "Error"; // cannot be true
```

The problem disappears if we use **$derived.by()**, has TypeScript understand that this code could be called later :
```ts
    let status: TileStatus = $state("Loading");
    const hasError = $derived.by(() => status == "Error"); // fine
```

**svelte2tsx** keeps `$derived()/$derived.by()` as is, but it should replace `$derived(X)` with `$derived.by(()=>X)`

**Comment by jycouet** [comment]
I'm not sure how should we proceed with this issue ? Move to cli ? Close ?

Seems that a lot of people are migrating with no issue. Because it's JS migration maybe ?
In one of my project, I have 280 errors after migration, I can of course édit manually my files.
Let me know

**Comment by brunnerh** [comment]
The migration should be fine, I think this belongs in [`language-tools`](https://github.com/sveltejs/language-tools); it's a matter of stopping the IDE from complaining.

**Comment by dummdidumm** [comment]
I am torn - since we can make this transformation only in Svelte files, you could have it working in one case but not the other and that could be very confusing. It's also nice that we don't need to transform the script tag, thought that's not as big of an argument

**Comment by marekdedic** [comment]
Hi,
I just accidentally opened a duplicate issue under svelte-check, partly because this one is rather hard to find and partly because IMO the underlying issue *has nothing to do with migrations*. The following code is something I can imagine anyone writing in Svelte 5:

```svelte
<script lang="ts">
  let one: Array<number> | null = $state(null);
  let two = $derived(
    one === null ? 42 : one.find((x) => x > 22)
  );
</script>
```
